rspace/src/main/scala/coop/rchain/rspace/history/Store.scala

```diff
@@ -31,98 +30,60 @@ final case class StoreConfig(
-      ret
-    } catch {
-      case NonFatal(ex) =>
-        ex.printStackTrace()
-        throw ex
-    } finally {
-      txn.close()
-    }
-
-  private def withReadTxnF[R](f: Txn[ByteBuffer] => R): F[R] =
-    Sync[F].delay {
-      withTxn(env.txnRead)(f)
-    }
-
-  private def withWriteTxnF[R](f: Txn[ByteBuffer] => R): F[R] =
-    Sync[F].delay {
-      withTxn(env.txnWrite)(f)
-    }
-
-  def get(key: Blake2b256Hash): F[Option[BitVector]] = {
-    val directKey = key.bytes.toDirectByteBuffer
-    get(directKey).map(v => v.map(BitVector(_)))
-  }
-
-  override def get(key: ByteBuffer): F[Option[ByteBuffer]] =
-    withReadTxnF { txn =>
-      dbi.get(txn, key)
-    }.map(v => Option(v))
-
-  def put(key: Blake2b256Hash, value: BitVector): F[Unit] = {
-    val directKey   = key.bytes.toDirectByteBuffer
-    val directValue = value.toByteVector.toDirectByteBuffer
-    put(directKey, directValue)
-  }
+  def lmdbStore[F[_]: Sync](config: StoreConfig): F[Store[F]] =
+    for {
+      env <- Sync[F].delay {
+              Env
+                .create()
+                .setMapSize(config.mapSize)
+                .setMaxDbs(config.maxDbs)
+                .setMaxReaders(config.maxReaders)
+                .open(config.path.toFile, config.flags: _*)
+            }
+      dbi   <- Sync[F].delay { env.openDbi("db", MDB_CREATE) }
+      store = LMDBStore(env, dbi)
+    } yield new Store[F] {
+      override def get(key: Blake2b256Hash): F[Option[BitVector]] = {
+        val directKey = key.bytes.toDirectByteBuffer
+        get(directKey).map(v => v.map(BitVector(_)))
+      }
 
-  @SuppressWarnings(Array("org.wartremover.warts.Throw"))
-  override def put(key: ByteBuffer, value: ByteBuffer): F[Unit] =
-    withWriteTxnF { txn =>
-      if (dbi.put(txn, key, value)) {
-        ()
-      } else {
-        throw new RuntimeException("was not able to put data")
+      override def put(key: Blake2b256Hash, value: BitVector): F[Unit] = {
+        val directKey   = key.bytes.toDirectByteBuffer
+        val directValue = value.toByteVector.toDirectByteBuffer
+        put(directKey, directValue)
       }
-    }
 
-  @SuppressWarnings(Array("org.wartremover.warts.Throw"))
-  private[this] def putIfAbsent(txn: Txn[ByteBuffer], key: ByteBuffer, value: ByteBuffer): Unit =
-    Option(dbi.get(txn, key)) match {
-      case None =>
-        if (dbi.put(txn, key, value)) {
-          ()
-        } else {
-          throw new RuntimeException("was not able to put data")
+      override def get(key: ByteBuffer): F[Option[ByteBuffer]] = store.get(key)
+
+      override def put(key: ByteBuffer, value: ByteBuffer): F[Unit] = store.put(key, value)
+
+      @SuppressWarnings(Array("org.wartremover.warts.Throw"))
+      private[this] def putIfAbsent(
+          txn: Txn[ByteBuffer],
+          key: ByteBuffer,
+          value: ByteBuffer
+      ): Unit =
+        Option(dbi.get(txn, key)) match {
+          case None =>
+            if (dbi.put(txn, key, value)) {
```

Comment:
This code looks similar to what's already in `LMDBStore`. Why not re-use?

Reply:
The code in `LMDBStore` encapsulates transactions inside of its methods. This method, on the other hand, is being used in rspace-specific `put` which runs multiple puts during a single transaction.